scripts/lib/CIME/XML/generic_xml.py

```diff
@@ -35,31 +57,169 @@ def __init__(self, infile=None, schema=None):
+        return node.xml_element.get(attrib_name, default=default)
+
+    def has(self, node, attrib_name):
+        return attrib_name in node.xml_element.attrib
+
+    def set(self, node, attrib_name, value):
+        return node.xml_element.set(attrib_name, value)
+
+    def pop(self, node, attrib_name):
+        return node.xml_element.attrib.pop(attrib_name)
+
+    def attrib(self, node):
+        # Return a COPY. We do not want clients making changes directly
+        return None if node.xml_element.attrib is None else dict(node.xml_element.attrib)
+
+    def set_name(self, node, name):
+        node.xml_element.tag = name
+
+    def set_text(self, node, text):
+        node.xml_element.text = text
+
+    def name(self, node):
+        return node.xml_element.tag
+
+    def text(self, node):
+        return node.xml_element.text
+
+    def add_child(self, node, root=None):
+        """
+        Add element node to self at root
+        """
+        root = root if root is not None else self.root
+        root.xml_element.append(node.xml_element)
+
+    def copy(self, node):
+        return deepcopy(node)
+
+    def remove_child(self, node, root=None):
+        root = root if root is not None else self.root
+        root.xml_element.remove(node.xml_element)
+
+    def make_child(self, name, attributes=None, root=None, text=None):
+        root = root if root is not None else self.root
+        if attributes is None:
+            node = _Element(ET.SubElement(root.xml_element, name))
+        else:
+            node = _Element(ET.SubElement(root.xml_element, name, attrib=attributes))
+
+        if text:
+            self.set_text(node, text)
+
+        return node
+
+    def get_children(self, name=None, attributes=None, root=None, no_validate=False):
+        """
+        This is the critical function, its interface and performance are crucial.
+
+        You can specify attributes={key:None} if you want to select chilren
+        with the key attribute but you don't care what its value is.
+        """
+        root = root if root is not None else self.root
+        children = []
+        for child in root.xml_element:
+            if name is not None:
+                if child.tag != name:
+                    continue
+
+            if attributes is not None:
+                if child.attrib is None:
+                    continue
+                else:
+                    match = True
+                    for key, value in attributes.iteritems():
+                        if key not in child.attrib:
+                            match = False
+                            break
+                        elif value is not None:
+                            if child.attrib[key] != value:
+                                match = False
+                                break
+
+                    if not match:
+                        continue
+
+            children.append(_Element(child))
+
+        # Remove
+        if not no_validate:
```

Comment:
Important: The global replacement of get_nodes with get_children means that all the calls that used to do recursive searches now only do direct-child searches. This validation code caught all the places that really needed the recursive scan and those places have since been changed to use scan_child or refactored so that the recursion is no longer needed.

This will incur a big performance hit. We should turn this off once we're confident in this big API refactor.